diff --git a/node_modules/mafmt/src/index.js b/node_modules/mafmt/src/index.js
index f2f8531..8a451fa 100644
--- a/node_modules/mafmt/src/index.js
+++ b/node_modules/mafmt/src/index.js
@@ -27,6 +27,11 @@ const WebSockets = or(
   and(DNS, base('ws'))
 )
 
+const TLS = or(
+  and(TCP, base('tls')),
+  and(DNS, base('tls'))
+)
+
 const WebSocketsSecure = or(
   and(TCP, base('wss')),
   and(DNS, base('wss'))
@@ -68,6 +73,7 @@ const Reliable = or(
   WebSocketsSecure,
   HTTP,
   HTTPS,
+  TLS
   WebRTCStar,
   WebRTCDirect,
   TCP,
@@ -129,6 +135,7 @@ exports.Stardust = Stardust
 exports.Circuit = Circuit
 exports.P2P = P2P
 exports.IPFS = P2P
+exports.TLS = TLS
 
 /*
  * Validation funcs
